plonk-hashing/src/poseidon/poseidon_ref.rs

```diff
@@ -0,0 +1,525 @@
+
+        inputs.iter().for_each(|x| {
+            let _ = poseidon.input(*x).unwrap();
+        });
+        let native_hash: Fr = poseidon.output_hash(&mut ());
+
+        let mut cs = ConstraintSystem::new_ref();
+        let mut poseidon_var =
+            PoseidonRef::<_, R1csSpecRef<Fr, WIDTH>, WIDTH>::new(
+                &mut cs,
+                param.clone(),
+            );
+        let inputs_var = inputs
+            .iter()
+            .map(|x| R1csSpecRef::<_, WIDTH>::alloc(&mut cs, *x))
+            .collect::<Vec<_>>();
+        inputs_var.iter().for_each(|x| {
+            let _ = poseidon_var.input(x.clone()).unwrap();
+        });
+
+        let hash_var = poseidon_var.output_hash(&mut cs);
+
+        assert!(cs.is_satisfied().unwrap());
+        assert_eq!(hash_var.value().unwrap(), native_hash);
+        println!(
+            "circuit size for WIDTH {} r1cs: {}",
+            WIDTH,
+            cs.num_constraints()
+        )
+    }
+
+    #[test]
+    #[should_panic]
+    // poseidon should output something if num_inputs > arity
+    fn sanity_test_failure() {
+        const ARITY: usize = 4;
+        const WIDTH: usize = ARITY + 1;
+        let mut rng = test_rng();
+
+        let param = PoseidonConstants::generate::<WIDTH>();
+        let mut poseidon =
+            PoseidonRef::<(), NativeSpecRef<Fr>, WIDTH>::new(&mut (), param);
+        (0..(ARITY + 1)).for_each(|_| {
+            let _ = poseidon.input(Fr::rand(&mut rng)).unwrap();
+        });
+        let _ = poseidon.output_hash(&mut ());
+    }
+
+    use crate::tests::{
+        conversion::cast_field,
+        neptune_hyper_parameter::collect_neptune_constants,
+    };
+    use neptune::{
+        poseidon::{HashMode, PoseidonConstants as NeptunePoseidonConstants},
+        Strength,
+    };
+    // let constants = NeptunePoseidonConstants::<Fr,
+    // A>::new_with_strength(strength); let mut p = NeptunePoseidon::<Fr,
+    // A>::new(&constants); let mut p2 = NeptunePoseidon::<Fr,
+    // A>::new(&constants); let mut p3 = NeptunePoseidon::<Fr,
+    // A>::new(&constants); let mut p4 = NeptunePoseidon::<Fr,
+    // A>::new(&constants);
+
+    //     let test_arity = constants.arity();
+    //     for n in 0..test_arity {
+    //         let scalar = Fr::from(n as u64);
+    //         p.input(scalar).unwrap();
+    //         p2.input(scalar).unwrap();
+    //         p3.input(scalar).unwrap();
+    //         p4.input(scalar).unwrap();
+    //     }
+
+    //     let digest = p.hash();
+    //     let digest2 = p2.hash_in_mode(Correct);
+    //     let digest3 = p3.hash_in_mode(OptimizedStatic);
+    //     let digest4 = p4.hash_in_mode(OptimizedDynamic);
+
+    #[test]
+    fn compare_with_neptune() {
+        const ARITY: usize = 2;
+        const WIDTH: usize = ARITY + 1;
+        type NepArity = generic_array::typenum::U2;
+
+        let (nep_consts, ark_consts) =
+            collect_neptune_constants::<NepArity>(Strength::Standard);
```

Comment:
`collect_neptune_constants` is missing, as well as neptune dev_dependencies? Nice to have these tests.

Reply:
This code is from `plonk_prototype` repo and we temporarily used neptune library as a reference to check the correctness of our code. 

Anyway, we should not even import `neptune` library here. We need to hardcoded tests.

Reply:
Those hardcoded test vectors are now here: https://github.com/Manta-Network/Plonk-Prototype/blob/d2e25aae7ad4526791f4543cebf1e4df5fbf14ac/src/poseidon/poseidon_ref.rs#L380-L415. Those test vectors come from my modified variants of the original sage script. @GhostOfGauss You can add this test here  remove any dependency on Neptune